### PR TITLE
[WIP] Basic rewrite of the package with addition of DSVI

### PR DESCRIPTION
**Pull request by theogf**
Hey, following #24 I am making a first attempt to make large changes (I don't believe this could be done by small changes and I don't think anyone is using this package atm anyway).

Here are the current main changes : 
- Introduction of a collection of variational distributions with specific parametrizations. This make it possible to change the variational parameters **in-place** and avoid passing a vector.
- The problem of having to deal with a **vector of variational parameters is gone** since by using the reparametrization trick, we take the gradients of the samples and not of the parameters.
- samples (and parametrized samples) are stored in so there is no additional allocation
- `optimize!` is now a loop over `step!` which has to be defined for each algorithm 
- Instead of `grad!`, there is now `gradlogπ!` and eventually `gradentropy`. The first is computed via sampling, the second mostly has closed-form solutions but a generic fall back with samples can be made.
- I entirely removed the `VariationalObjective` object which right now objectively does not bring anything (I guess if we start to consider more divergences we could start to add such objects)
- I removed Tracker to put it as a `Requires` based package (does anyone still uses Tracker?)
- More things to come

### CODE REVIEW

**Comment by yebai** [comment]
thanks @theogf, it looks like a great PR in progress. Just a clarification question, does `DSVI` = `ADVI`?

**Comment by theogf** [comment]
> thanks @theogf, it looks like a great PR in progress. Just a clarification question, does `DSVI` = `ADVI`?

That's a good question, after checking the ADVI paper, ADVI is DSVI with bijectors (which is answering the question I had @torfjelde). However the previous implementation of ADVI was not reflecting what was done in the paper. 

Reference: Automatic Differentiation Variational Inference https://arxiv.org/abs/1603.00788

**Comment by theogf** [comment]
Here is a larger discussion on having `q(theta)` vs/with `q`.

So I implemented `ADVI` and a basic version of `BBVI`. Interestingly, the first one relies on `q` and the second one on `q(theta)`.
This leads me to think that the choice `q(theta)`/`q` literally depends of the algorithm used and should be treated as such.

The approach I took now for `BBVI` is to use the amazing `state` approach. Since we work with our own distributions (but this is not a restriction), I defined a `to_vec(q)` and `to_dist(q, theta)` to jump between the two representations. This way we create an initial `theta` in `init` that we update, and finally return `to_dist(q, theta)` at the end of the run.

So basically I would argue to only leave the `q` approach and to eventually deal with the `q(theta)` internally.

**Comment by theogf** [comment]
Since (most) tests are passing I am going to make it ready for review while I work on the tests and remove some unnecessary bits.

One part I am quite unhappy about is the way gradients are computed. Every algorithm require a different approach and I could not find a global approach. This means a huge load of code copy/paste to make it work with the different backends...
I would be happy to hear some suggestions

**Comment by theogf** [comment]
I solved the gradient issue with going back to `grad!` as a very generic inplace gradient computing function.
Tests are sometimes not passing because of their randomness, I don't know if there is a better approach here.

File: src/algorithms/bbvi.jl [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+$(TYPEDEF)
+
+Black-Box Variational Inference (BBVI) with automatic differentiation
+backend `AD`.
+
+# Fields
+
+$(TYPEDFIELDS)
+"""
+struct BBVI{AD} <: VariationalInference{AD}
```
Comment (luiarthur) [inline src/algorithms/bbvi.jl:11]: It's awesome to see BBVI here. I didn't see any implementations for rao-blackwellization / control-variates to reduce variance of the noisy gradient. But I suppose this could be extended.

File: src/algorithms/bbvi.jl [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    rand!(q̂, state.x) # Get initial samples from x₀
+    Δlog = evaluate.(logπ, Ref(q̂), eachcol(state.x)) .- logpdf(q̂, state.x)
+    f(θ) = dot(logpdf(to_dist(q, θ), state.x), Δlog) / nsamples(alg)
+    grad!(state.diff_result, f, state.θ, alg)
+    return update!(alg, q, state, opt)
+end
+
+function update!(::BBVI, q, state, opt)
+    state.θ .+= Optimise.apply!(opt, state.θ, DiffResults.gradient(state.diff_result))
+    return nothing
```
Comment (luiarthur) [inline src/algorithms/bbvi.jl:54]: Isn't the convention to return the mutated state? Any reason for returning `nothing` instead of `state.θ`?

File: src/algorithms/bbvi.jl (hunk 1 above)
Comment (theogf) [reply src/algorithms/bbvi.jl:11]: That's indeed my intention, I just thought it would be good to have at least two VI implementations! Control variates and rao Blackwell will follow :)

File: src/algorithms/bbvi.jl (hunk 2 above)
Comment (theogf) [reply src/algorithms/bbvi.jl:54]: Yeah that's a good point, I am not sure what should be the right thing to do here (I will ask on Slack).

File: src/distributions/distributions.jl [hunk 3]
```diff
@@ -0,0 +1,58 @@
+## Series of variation of the MvNormal distribution, different methods need different parametrizations ##
+abstract type AbstractPosteriorMvNormal{T} <:
+              Distributions.ContinuousMultivariateDistribution end
```
Comment (torfjelde) [inline src/distributions/distributions.jl:3]: Why is this needed vs. `AbstractMvNormal` or whatever it's called?

File: src/distributions/cholmvnormal.jl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+
+function reparametrize!(x, q::CholMvNormal, z)
+    x .= q.μ .+ q.Γ * z
+end
+
+
+function to_vec(q::CholMvNormal)
+    vcat(q.μ, vec(q.Γ))
+end
```
Comment (torfjelde) [inline src/distributions/cholmvnormal.jl:35]: Can we not make use of something similar to `Flux.params` here since we already depend on `Functors.jl`?

**Review by torfjelde — CHANGES_REQUESTED** [review]
Good stuff! There's quite a bit that needs addressing though.

I've left some comments on what for sure needs changing. I still need to look through again and think about how we can get to a complete solution here.

Some more general comments:
1. I would really, really like to not reimplement a bunch of distributions in this package. Most of these very particular distributions could probably go into DistributionsAD.jl if we need them.
2. There's a big focus on mutating operations here. It seems a bit unnecessary given that it's a matter of decreasing memory usage by a constant factor of 2, no? Or am I missing something here? Also, you've mentioned Optimisers.jl; isn't that moving towards *non-mutation* states for optimisers? 


But yeah, after you've responded to my comments, I'll have to go through again :+1:

File: src/algorithms/advi.jl [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        return evaluate(logπ, q, x)
+    end
+    grad!(state.diff_result, f, state.x, alg)
+    return update!(alg, q, state, opt)
+end
+
+function update!(alg::ADVI, q, state, opt)
+    Δ = DiffResults.gradient(state.diff_result)
+    update_mean!(q, vec(mean(Δ, dims = 2)), opt)
+    update_cov!(alg, q, Δ, state, opt)
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/algorithms/advi.jl:56]: Well actually not, and that's where I am trying to be maybe more precise. The ADVI approach I implemented is following the ADVI paper which namely states : 
image
So the underlying distribution is always a Gaussian.
I will add a reference to the paper in the docs

If we want to train something different it will have to be a different algorithm.

File: src/algorithms/advi.jl [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+function ADVI(samples_per_step::Int = 1, max_iters::Int = 1000)
+    return ADVI{ADBackend()}(samples_per_step, max_iters)
+end
+
+alg_str(::ADVI) = "ADVI"
+nsamples(alg::ADVI) = alg.samples_per_step
+niters(alg::ADVI) = alg.max_iters
+
+function compats(::ADVI)
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/algorithms/advi.jl:26]: See the previous point.

File: src/algorithms/bbvi.jl [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+end
+
+function finish(::BBVI, q, state)
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/algorithms/bbvi.jl:57]: Hahaha, ok not the best name in retrospective. The idea is to perform some post processing step at the end. I will rename it `final_dist` for final distination

File: src/algorithms/advi.jl [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    samples_per_step = nsamples(alg)
+    x₀ = rand(q, samples_per_step) # Preallocating x₀
+    x = similar(x₀) # Preallocating x
+    diff_result = DiffResults.GradientResult(x)
+    return (x₀=x₀, x=x, diff_result=diff_result)
+end
+
+function step!(::ELBO, alg::ADVI, q, logπ, state, opt)
+    randn!(state.x₀) # Get initial samples from x₀
+    reparametrize!(state.x, q, state.x₀)
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/algorithms/advi.jl:45]: So you take `x_0` some nice normal samples, you reparametrize them and then you compute the gradients of the logjoint given these samples and do some more reparametrization magic (see the ADVI paper) and get the gradients out!

File: src/interface.jl [hunk 9]
```diff
@@ -0,0 +1,85 @@
+"""
+    vi([vo, [model, alg::VariationalInference]]; opt, hyperparams, opt_hyperparams)
+    vi([vo, [model, alg::VariationalInference, q::VariationalPosterior]]; opt, hyperparams, opt_hyperparams)
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/interface.jl:3]: Yeah not really I changed it to something standard now

File: src/interface.jl [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+function optimize!(
+    vo::VariationalObjective,
+    alg::VariationalInference,
+    q,
+    model;
+    opt = TruncatedADAGrad(),
+    hyperparams = nothing,
+    opt_hyperparams = nothing,
+)
+    # TODO: should we always assume `samples_per_step` and `max_iters` for all algos?
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/interface.jl:42]: That was leftover from before :)

File: src/interface.jl [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+    max_iters = niters(alg)
+    
+    global state = init(alg, q, opt) # opt is there to be used in the future
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/interface.jl:45]: Debugging leftovers :)

File: src/interface.jl [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    end
+end
+
+function compat(alg::VariationalInference, q)
+    return q isa compats(alg)
+end
+
+function compats(::Any)
+    return ()
+end
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/interface.jl:85]: I think my solution is awesome :D

File: src/utils.jl [hunk 13]
```diff
@@ -0,0 +1,19 @@
+XXt(A) = A * A'
+XtX(A) = A' * A
```
[the remark this answers is not in the document]
Comment (theogf) [reply src/utils.jl:2]: This avoid this kind of computations:
```
A = awesome_computations
W = A * A'
```
or alternatively
```
W = A |> x -> A * A'
```
Which are happening a lot.

File: src/distributions/distributions.jl (hunk 3 above)
Comment (theogf) [reply src/distributions/distributions.jl:3]: I was not aware of `AbstractMvNormal` until recently. This is still practical to have `AbstractPosteriorMvNormal` to define stuff like `mean(d::AbstractPosteriorMvNormal) = d.\mu`

File: src/distributions/cholmvnormal.jl (hunk 4 above)
Comment (theogf) [reply src/distributions/cholmvnormal.jl:35]: Well you need to tell that to ForwardDiff.jl, ReverseDiff.jl and co :D

File: test/runtests.jl [hunk 14]
```diff
@@ -1,28 +1,27 @@
...
+        include(joinpath("algorithms", "bbvi.jl"))
+    end
+    @testset "distributions" begin
+        include(joinpath("distributions", "distributions.jl"))
+        include(joinpath("distributions", "diagmvnormal.jl"))
+        include(joinpath("distributions", "cholmvnormal.jl"))
+    end
+    include("gradients.jl")
+    include("interface.jl")
+    # include("optimisers.jl") # Relying on Tracker...
```
[the remark this answers is not in the document]
Comment (theogf) [reply test/runtests.jl:24]: I think this is currently outdated, I need to have another look

[the remark this answers is not in the document]
Comment (theogf) [reply test/runtests.jl:24]: Yeah the problem is that it only works with Tracker and I had deleted it for a while

File: src/algorithms/advi.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (torfjelde) [reply src/algorithms/advi.jl:60]: I'm fully aware that the original paper only had this in mind, but it's unecessarily restricted and is mainly just a consequence of the restricted tools they had available at the time.

It's called "Automatic Differential Variational Inference" which really refers to the fact that we use AD + reparameterization trick.

IMO it just comes down to: what do it cost us to allow *any* reparameterization with *any* valid base distribution, and just let Gaussian be the default? That just seems overall superior, no?

File: src/distributions/cholmvnormal.jl (hunk 4 above)
Comment (torfjelde) [reply src/distributions/cholmvnormal.jl:36]: There are no ties between `Flux.params` and the AD-framework used anymore:)
It's using Functors.jl to define what's trainable and what isn't. So `Flux.params` should *just work* for the above distributions. Then you just `vec` and `vcat`.

Comment (theogf) [reply src/distributions/cholmvnormal.jl:36]: Ah sorry I misunderstood you, I thought you were complaining about the `vcat`.

File: src/algorithms/advi.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (theogf) [reply src/algorithms/advi.jl:60]: Well in theory yes that sounds amazing. The problem comes from how do you formulate the reparametrization trick for non-Gaussian. It's not always possible right?

File: src/utils.jl (hunk 13 above)
[the remark this answers is not in the document]
Comment (torfjelde) [reply src/utils.jl:2]: Sure but it makes the code unnecesarily complicated to read though? And we're talking about making
```julia
A = awesome_computations
W = A * A'
```
into
```julia
W = XXt(awesome_computations)
```
?
You can also use `let` statements if you don't want the variable to be in the main scope:
```julia
W = let A = awesome_computations
    A * A'
end
```

IMO it makes the code less readable, but I'm not too fussed about it so feel free to leave it as is :+1:

File: test/runtests.jl (hunk 14 above)
[the remark this answers is not in the document]
Comment (torfjelde) [reply test/runtests.jl:24]: Am confused. Isn't using ForwardDiff? o.O

File: src/utils.jl (hunk 13 above)
[the remark this answers is not in the document]
Comment (theogf) [reply src/utils.jl:2]: Ok yeah the `let` option is pretty cool, I am just not used to it.

File: src/algorithms/advi.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (torfjelde) [reply src/algorithms/advi.jl:60]: Normalizing flows; affine transformation is just an instance:)

File: test/runtests.jl (hunk 14 above)
[the remark this answers is not in the document]
Comment (theogf) [reply test/runtests.jl:24]: See here : https://github.com/TuringLang/AdvancedVI.jl/blob/808acbe46d68ecb8ce89d18a4470b1c162feb2cb/src/optimisers.jl#L42

File: src/algorithms/advi.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (torfjelde) [reply src/algorithms/advi.jl:60]: IMO, we don't even need all these implementations of differnt Gaussians. We just make it different affine transformations, e.g. introduce a `Affine <: Bijector` and define different evaluations depending on whether the scaling-matrix is `Cholesky`, etc.

[the remark this answers is not in the document]
Comment (torfjelde) [reply src/algorithms/advi.jl:60]: The above is the reasoning behind the current impl btw. Though I 100% agree we should make it more convenient to instantiate the different approaches.

File: src/algorithms/advi.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (torfjelde) [reply src/algorithms/advi.jl:49]: But where's the `logabsdetjac` term?

**Comment by trappmartin** [comment]
Is there a timeframe for this PR?

**Comment by theogf** [comment]
> Is there a timeframe for this PR?

Not really, I think the problem is that it tries to put too many things at the same time.
Also there are some things where we are not entirely sure on how to proceed.

@torfjelde What do you think about having the variational distributions defined as bijected distributions in another PR?
